skipper/userinfo/userinfo.go

```diff
@@ -0,0 +1,81 @@
+package userinfo
+
+import (
+	"fmt"
+
+	authenticationv1 "k8s.io/api/authentication/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	labels "k8s.io/apimachinery/pkg/labels"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+)
+
+const (
+	// saPrefix represents the service account prefix in admission requests
+	saPrefix = "system:serviceaccount:"
+
+	clusterRoleKind = "ClusterRole"
+	roleKind        = "Role"
+)
+
+// RoleRefs gets the list of roles and cluster roles for the incoming api-request
+func RoleRefs(rbLister rbacv1listers.RoleBindingLister, crbLister rbacv1listers.ClusterRoleBindingLister, user authenticationv1.UserInfo) (roles []string, clusterroles []string, err error) {
+	roleBindings, err := rbLister.List(labels.Everything())
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to list rolebindings: %v", err)
+	}
+	rs, crs := roleRefs(roleBindings, user)
+
+	clusterroleBindings, err := crbLister.List(labels.NewSelector())
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to list clusterrolebindings: %v", err)
+	}
+	crs = append(crs, clusterRoleRefs(clusterroleBindings, user)...)
+	return rs, crs, nil
+}
+
+func roleRefs(roleBindings []*rbacv1.RoleBinding, userInfo authenticationv1.UserInfo) (roles []string, clusterRoles []string) {
+	for _, rolebinding := range roleBindings {
+		for _, subject := range rolebinding.Subjects {
+			if matchSubject(subject, userInfo, rolebinding.Namespace) {
+				switch rolebinding.RoleRef.Kind {
+				case roleKind:
+					roles = append(roles, rolebinding.Namespace+":"+rolebinding.RoleRef.Name)
+				case clusterRoleKind:
+					clusterRoles = append(clusterRoles, rolebinding.RoleRef.Name)
+				}
+			}
+		}
+	}
+	return roles, clusterRoles
+}
+
+func clusterRoleRefs(clusterroleBindings []*rbacv1.ClusterRoleBinding, userInfo authenticationv1.UserInfo) (clusterRoles []string) {
+	for _, clusterRoleBinding := range clusterroleBindings {
+		for _, subject := range clusterRoleBinding.Subjects {
+			if clusterRoleBinding.RoleRef.Kind == clusterRoleKind {
+				if matchSubject(subject, userInfo, subject.Namespace) {
+					clusterRoles = append(clusterRoles, clusterRoleBinding.RoleRef.Name)
+				}
+			}
+		}
+	}
+	return clusterRoles
+}
+
```

Comment:
I mean I hate this, but I don't have a better idea. I guess this is also the way kyverno does it?

Reply:
I very much hate it too. Yes Kyverno does it almost in the same way.

Reply:
Ok, then I'd leave it as is. There is probably room for optimization, but I don't think we actually need it